nucypher/blockchain/eth/actors.py

```diff
@@ -666,6 +515,127 @@ def set_min_reward_rate_range(self,
+            self._add_substake(staker, amount, lock_periods)
+            total_to_allocate += amount
+
+        token_agent = ContractAgency.get_agent(NucypherTokenAgent, registry=registry)  # type: NucypherTokenAgent
+
+        balance = token_agent.get_balance(deployer_address)
+        if balance < total_to_allocate:
+            raise ValueError(f"Not enough tokens to allocate. We need at least {NU.from_nunits(total_to_allocate)}.")
+
+        allowance = token_agent.get_allowance(owner=deployer_address, spender=self.staking_agent.contract_address)
+        if allowance < total_to_allocate:
+            self.log.debug(f"Allocating a total of {NU.from_nunits(total_to_allocate)}")
+            allowance_function = token_agent.contract.functions.increaseAllowance(self.staking_agent.contract_address,
+                                                                                  total_to_allocate - allowance)
+            _allowance_receipt = self.staking_agent.blockchain.send_transaction(contract_function=allowance_function,
+                                                                                sender_address=deployer_address)
+
+    def _add_substake(self, staker, amount, lock_periods):
+        try:
+            substakes = self.allocations[staker]
+            if len(substakes) >= self.max_substakes:
+                raise ValueError
+        except KeyError:
+            if list(self.staking_agent.get_all_stakes(staker_address=staker)):
+                raise ValueError(f"{staker} is already a staker. It cannot be included in a batch deposit")
+            substakes = list()
+            self.allocations[staker] = substakes
+
+        message = f"Invalid substake for {staker}: "
+        if amount < self.economics.minimum_allowed_locked:
+            message += f"Amount ({amount}) is below the min allowed ({self.economics.minimum_allowed_locked})"
+            raise ValueError(message)
+        overall_amount = sum([amount for amount, periods in substakes])
+        if overall_amount + amount > self.economics.maximum_allowed_locked:
+            message += f"Total amount is above the max allowed ({self.economics.maximum_allowed_locked})"
+            raise ValueError(message)
+        if lock_periods < self.economics.minimum_locked_periods:
+            message += f"Lock periods ({lock_periods}) are below the min ({self.economics.minimum_locked_periods})"
+            raise ValueError(message)
+
+        substakes.append((amount, lock_periods))
+
+    def deposit_next_batch(self,
+                           sender_address: str,
+                           gas_limit: int = None) -> Tuple[List[str], dict]:
+
+        pending_stakers = [staker for staker in self.allocations.keys() if staker not in self.deposited]
+
+        self.log.debug(f"Constructing next batch. "
+                       f"Currently, {len(pending_stakers)} stakers pending, {len(self.deposited)} deposited.")
+
+        batch_size = 1
```

Comment:
What's the average batch size that we end up with? Do we expect the batch size to vary widely between batches? From an efficiency perspective, I'm wondering if it makes sense to also store/provide the last good batch size from a previous call, and then start from there, and if that fails decrement down, instead of starting from 1 and incrementing every time.

Reply:
The idea makes sense, but the code would be probably more complex, as it'd have to keep some state between batches, and also I dont know how it can handle different types of stakers (i.e., variation on the number of substakes). In practice, the allocations we're going to do will probably allow for less than 10 stakers per batch, so starting from 1 is not that big deal.

Reply:
👍 - Fair point although I think `batch_deposits` could pass the last batch size to `deposit_next_batch`. Not a huge issue though 😄